### ISSUE TITLE
fails to load chain if any onion peers connected

### ISSUE THREAD

**Issue by tril0byte** [dgarage/NBXplorer#56]
``info: Configuration:  BTC: Loading chain from node... 
fail: Configuration:  BTC: Unhandled exception in BitcoinDWaiter 
System.OperationCanceledException: Loading the chain from the node timed out 
   at NBXplorer.BitcoinDWaiter.<ConnectToBitcoinD>d__36.MoveNext() in /path/to/NBxplorer/NBXplorer/BitcoinDWaiter.cs:line 349 
--- End of stack trace from previous location where exception was thrown --- 
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() 
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
   at NBXplorer.BitcoinDWaiter.<StepAsync>d__35.MoveNext() in /path/to/NBxplorer/NBXplorer/BitcoinDWaiter.cs:line 263 
--- End of stack trace from previous location where exception was thrown --- 
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() 
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult() 
   at NBXplorer.BitcoinDWaiter.<StartLoop>d__27.MoveNext() in /path/to/NBxplorer/NBXplorer/BitcoinDWaiter.cs:line 183 

I believe this is a bug in underlying NBitcoin library. However since I do not have the NBitcoin installed and am using it through NBXplorer, which downloaded NBitcoin for me into my .nuget directory, maybe you should check what version of NBitcoin you are using and possibly update, or report more details upstream. NBitcoin appears to run a getpeerinfo and then immediately throw the cancelled exception if any onion peers are connected to the bitcoin node. It's not a timeout, the error is misleading. Reproduce by running a bitcoin node with -onion=127.0.0.1:9050 and -onlynet=onion and wait for a peer and then run NBxplorer.

**Comment by NicolasDorier** [comment]
NBXplorer can't connect if you use `-onlynet=tor` as NBitcoin does not support TOR and assume you are connected on the same machine.

Try to add `whitelist=127.0.0.1` to your configuration file of bitcoin.

**Comment by tril0byte** [comment]
I only said onlynet to reproduce.  The bug occurs if ANY onion peers are connected, even if other peers are too.

**Comment by tril0byte** [comment]
I do not plan to whitelist 127.0.0.1 because that will whitelist all tor peers.
I already whitelisted my NBXplorer on another port as well. This is an issue triggered with RPC only.

**Comment by NicolasDorier** [comment]
what about using `whitebind` and connect nbxplorer to this port? That way NBXplorer would be properly whitelisted, but not other TOR peers. I am quite sure the underlying reason is this.

**Comment by tril0byte** [comment]
Either I don't understand what you are saying or you don't follow my bug report.   I'm already doing exactly what you just suggested.   I have a additional whitebind for port 8330 and I told NBXplorer to use it - it removed a warning about whitelisting, and I verified it connect using the port.  I am not sure why you are focusing on the P2P port as I am not seeing how it is related to this at all.  NBXplorer starts up properly and connects and goes to Ready state - as long as there are no onion peers connected to the bitcoin node.  If any onion peers are active, NBXplorer prints the above exception and goes into a loop, trying to reconnect to bitcoin and getting disconnected (until the onion peers are all gone).  What other information can I provide?

Here are my config files, modified to remove specific identifiers.

~/.nbxplorer/Main/settings.config:
btc.node.endpoint=127.0.0.1:8330

~/.bitcoin/bitcoin.conf:
debug=rpc 
daemon=1 
listen=1 
onion=127.0.0.1:9050 
discover=0 
bind=<PUBLIC_IPv4>:8333 
bind=127.0.0.1:8333 
bind=<CJDNS_IPv6> 
bind=<HE.NET_IPv6> 
whitebind=127.0.0.1:8330 
externalip=<PUBLIC_IPv4> 
externalip=<hidden service .onion> 
externalip=<CJDNS_IPv6> 
externalip=<HE.NET_IPv6> 
maxconnections=500 
rpcallowip=127.0.0.1 
rpcallowip=::1 
rpctimeout=30 
disablewallet=1 
addnode=v4.us-west.fibre.bitcoinrelaynetwork.org 
whitelist=<IPv4_Trusted_Peer1> 
whitelist=<IPv4_Trusted_Peer2> 
whitelist=<CJDNS_Trusted_Peer>

**Comment by tril0byte** [comment]
here is the part of .bitcoin/debug.log that is generated by NBXplorer's RPC calls:

2018-05-06 18:47:35 ThreadRPCServer method=validateaddress
2018-05-06 18:47:35 ThreadRPCServer method=getnetworkinfo
2018-05-06 18:47:35 ThreadRPCServer method=getblockchaininfo
2018-05-06 18:47:38 ThreadRPCServer method=getpeerinfo

It always crashes right after the getpeerinfo.

Here is the full NBXplorer output - modified to remove identifiers:
$ dotnet run 
info: Configuration:  Data Directory: /path/to/.nbxplorer/Main 
info: Configuration:  Configuration File: /path/to/.nbxplorer/Main/settings.config 
info: Configuration:  Network: Mainnet 
info: Configuration:  Supported chains: BTC 
info: Configuration:  Network: Mainnet 
info: Configuration:  Supported chains: BTC 
info: Configuration:  BTC: Testing RPC connection to http://localhost:8332/ 
Hosting environment: Production 
Content root path: /path/to/NBxplorer/NBXplorer/bin/Debug/netcoreapp2.0/ 
Now listening on: http://127.0.0.1:24444 
Application started. Press Ctrl+C to shut down. 
info: Configuration:  BTC: RPC connection successfull 
info: Configuration:  BTC: Full node version detected: 160000 
info: Configuration:  BTC: Loading chain from cache... 
info: Configuration:  BTC: Height: 521430 
info: Configuration:  BTC: Loading chain from node... 
fail: Configuration:  BTC: Unhandled exception in BitcoinDWaiter 
System.OperationCanceledException: Loading the chain from the node timed out 
   at NBXplorer.BitcoinDWaiter.<ConnectToBitcoinD>d__36.MoveNext() in /path/to/NBxplorer/NBXplorer/BitcoinDWaiter.cs:line 349 
--- End of stack trace from previous location where exception was thrown --- 
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() 
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
   at NBXplorer.BitcoinDWaiter.<StepAsync>d__35.MoveNext() in /path/to/NBxplorer/NBXplorer/BitcoinDWaiter.cs:line 263 
--- End of stack trace from previous location where exception was thrown --- 
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() 
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult() 
   at NBXplorer.BitcoinDWaiter.<StartLoop>d__27.MoveNext() in /path/to/NBxplorer/NBXplorer/BitcoinDWaiter.cs:line 183 
^CApplication is shutting down...

**Comment by tril0byte** [comment]
if I don't hit ^C it goes back to 
info: Configuration:  BTC: Testing RPC connection [...] and repeats,
and the debug.log repeats all 4 lines each loop.

**Comment by tril0byte** [comment]
added -whitelist=127.0.0.1, which had no effect.

**Comment by NicolasDorier** [comment]
I think I know the issue does: `onlynet=tor` ensure that inbound connection are using tor? If so, this is a problem for NBXplorer which can't use tor.

**Comment by tril0byte** [comment]
I reproduced the issue by using addnode from another node to this node's onion address. I only said to use -onlynet because it makes outbound connections to onion and thus triggers the bug faster. I don't use onlynet in regular operation.

**Comment by NicolasDorier** [comment]
Trying to get TOR to work:

Given
```
$env:Path = $env:Path + ";C:\Users\NicolasDorier\Downloads\tor-win32-0.3.2.10\Tor"
$env:Path = $env:Path + ";C:\Sources\NBitcoin\NBitcoin.Tests\bin\Debug\net461\TestData\bitcoin-0.16.0\bin"
$dataDirBase = "C:\Sources\NBitcoin\NBitcoin.Tests\bin\Debug\net461\TestData\bitcoin-0.16.0\bin"
$nbxplorer = "C:\Sources\NBXplorer\NBXplorer"

mkdir "$dataDirBase\Instance1"
mkdir "$dataDirBase\Instance2"
.\tor.exe -controlport 9051 -cookieauthentication 1
.\bitcoind -regtest -printtoconsole -onion="127.0.0.1:9050" -bind="127.0.0.1:8335" -rpcport=8334 -debug=1 -datadir="$dataDirBase\Instance1"
.\bitcoind -regtest -printtoconsole -onion="127.0.0.1:9050" -bind="127.0.0.1:8336" -rpcport=8337 -connect="o5doi7b72ljhn6ak.onion:18444" -debug=1 -datadir="$dataDirBase\Instance2"

cd "$nbxplorer"
./run.ps1 --chains=btc -regtest --btcnodeendpoint="127.0.0.1:8335" --btcrpcurl="http://127.0.0.1:8334" --btcrpccookiefile="$dataDirBase\Instance1\regtest\.cookie"
```

Running this now. However, having problem to connect my two bitcoind instance together.  Trying to find out why.

**Comment by NicolasDorier** [comment]
Mmmh they don't want to connect to each other:

```
2018-05-09 03:04:05 init message: Loading P2P addresses...
2018-05-09 03:04:05 tor: Reading cached private key from C:\Sources\NBitcoin\NBitcoin.Tests\bin\Debug\net461\TestData\bitcoin-0.16.0\bin\Instance2\regtest\onion_private_key
2018-05-09 03:04:05 tor: Successfully connected!
2018-05-09 03:04:05 Loaded 0 addresses from peers.dat  1ms
2018-05-09 03:04:05 tor: Connected to Tor version 0.3.2.10
2018-05-09 03:04:05 init message: Loading banlist...
2018-05-09 03:04:05 tor: Supported authentication method: COOKIE
2018-05-09 03:04:05 Loaded 0 banned node ips/subnets from banlist.dat  2ms
2018-05-09 03:04:05 tor: Supported authentication method: SAFECOOKIE
2018-05-09 03:04:05 init message: Starting network threads...
2018-05-09 03:04:05 tor: Using SAFECOOKIE authentication, reading cookie authentication from C:\Users\NicolasDorier\AppData\Roaming\tor\control_auth_cookie
2018-05-09 03:04:05 DNS seeding disabled
2018-05-09 03:04:05 tor: SAFECOOKIE authentication challenge successful
2018-05-09 03:04:05 net thread start
2018-05-09 03:04:05 init message: Done loading
2018-05-09 03:04:05 tor: AUTHCHALLENGE ServerHash 05a9d714eedbc1d268a9f75627f66221ec7a55ab2a95838363860ed895874183 ServerNonce f4db11485ea25e9139e6541ff99bbe1c6673d7cfb76a531dbf13d877700f66b1
2018-05-09 03:04:05 opencon thread start
2018-05-09 03:04:05 addcon thread start
2018-05-09 03:04:05 msghand thread start
2018-05-09 03:04:05 tor: Authentication successful
2018-05-09 03:04:05 trying connection o5doi7b72ljhn6ak.onion:18444 lastseen=0.0hrs
2018-05-09 03:04:05 SOCKS5 connecting o5doi7b72ljhn6ak.onion
2018-05-09 03:04:05 tor: ADD_ONION successful
2018-05-09 03:04:05 SOCKS5 sending proxy authentication 0:0
2018-05-09 03:04:05 tor: Got service ID rwnsbkcfbzmpacaj, advertising service rwnsbkcfbzmpacaj.onion:18444
2018-05-09 03:04:05 tor: Cached service private key to C:\Sources\NBitcoin\NBitcoin.Tests\bin\Debug\net461\TestData\bitcoin-0.16.0\bin\Instance2\regtest\onion_private_key
2018-05-09 03:04:05 AddLocal(rwnsbkcfbzmpacaj.onion:18444,4)
2018-05-09 03:04:14 Socks5() connect to o5doi7b72ljhn6ak.onion:18444 failed: general failure
2018-05-09 03:04:15 trying connection o5doi7b72ljhn6ak.onion:18444 lastseen=0.0hrs
2018-05-09 03:04:15 SOCKS5 connecting o5doi7b72ljhn6ak.onion
2018-05-09 03:04:15 SOCKS5 sending proxy authentication 1:1
2018-05-09 03:04:22 Socks5() connect to o5doi7b72ljhn6ak.onion:18444 failed: general failure
2018-05-09 03:04:23 trying connection o5doi7b72ljhn6ak.onion:18444 lastseen=0.0hrs
2018-05-09 03:04:23 SOCKS5 connecting o5doi7b72ljhn6ak.onion
2018-05-09 03:04:23 SOCKS5 sending proxy authentication 2:2
2018-05-09 03:04:31 Socks5() connect to o5doi7b72ljhn6ak.onion:18444 failed: general failure
2018-05-09 03:04:32 trying connection o5doi7b72ljhn6ak.onion:18444 lastseen=0.0hrs
2018-05-09 03:04:32 SOCKS5 connecting o5doi7b72ljhn6ak.onion
2018-05-09 03:04:32 SOCKS5 sending proxy authentication 3:3
2018-05-09 03:04:39 Socks5() connect to o5doi7b72ljhn6ak.onion:18444 failed: general failure
2018-05-09 03:04:41 trying connection o5doi7b72ljhn6ak.onion:18444 lastseen=0.0hrs
2018-05-09 03:04:41 SOCKS5 connecting o5doi7b72ljhn6ak.onion
2018-05-09 03:04:41 SOCKS5 sending proxy authentication 4:4
2018-05-09 03:04:49 Socks5() connect to o5doi7b72ljhn6ak.onion:18444 failed: general failure
```

No error on TOR

```
PS C:\Users\NicolasDorier\Downloads\tor-win32-0.3.2.10\Tor> .\tor.exe -controlport 9051 -cookieauthentication 1
May 09 12:03:18.463 [notice] Tor 0.3.2.10 running on Windows 8 with Libevent 2.0.22-stable, OpenSSL 1.0.2n, Zlib 1.2.8, Liblzma N/A, and Libzstd N/A.
May 09 12:03:18.464 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://www.torproject.org/download/download#warning
May 09 12:03:18.495 [notice] Configuration file "C:\Users\NicolasDorier\AppData\Roaming\tor\torrc" not present, using reasonable defaults.
May 09 12:03:18.562 [warn] Path for GeoIPFile (<default>) is relative and will resolve to C:\Users\NicolasDorier\Downloads\tor-win32-0.3.2.10\Tor\<default>. Is this what you wanted?
May 09 12:03:18.563 [warn] Path for GeoIPv6File (<default>) is relative and will resolve to C:\Users\NicolasDorier\Downloads\tor-win32-0.3.2.10\Tor\<default>. Is this what you wanted?
May 09 12:03:18.572 [notice] Scheduler type KISTLite has been enabled.
May 09 12:03:18.576 [notice] Opening Socks listener on 127.0.0.1:9050
May 09 12:03:18.583 [notice] Opening Control listener on 127.0.0.1:9051
May 09 12:03:18.000 [notice] Bootstrapped 0%: Starting
May 09 12:03:20.000 [notice] Starting with guard context "default"
May 09 12:03:20.000 [notice] Bootstrapped 80%: Connecting to the Tor network
May 09 12:03:22.000 [notice] Bootstrapped 85%: Finishing handshake with first hop
May 09 12:03:23.000 [notice] Bootstrapped 90%: Establishing a Tor circuit
May 09 12:03:23.000 [notice] New control connection opened from 127.0.0.1.
May 09 12:03:24.000 [notice] Tor has successfully opened a circuit. Looks like client functionality is working.
May 09 12:03:24.000 [notice] Bootstrapped 100%: Done
May 09 12:04:05.000 [notice] New control connection opened from 127.0.0.1.
```

The Instance1 does not see the connection attempt.

**Comment by NicolasDorier** [comment]
Help me to get TOR work between 2 bitcoin instance. I can't get it to work properly.

**Comment by tril0byte** [comment]
hm, you seem to be on windows and using bitcoin to control the tor instance. Neither of which I am familiar with. Is the bitcoin instance accepting peer connections? Is this the same tor instance for 2 bitcoins? General failure doesn't give much detail. Can you try this tor instance to see if it is working, such as set browser socks5 proxy: port 9050 and  browse to https://check.torproject.org ?  Did you wait a while after starting the bitcoin instances?  When publishing an onion service it may take 10 minutes before another node can connect.

**Comment by NicolasDorier** [comment]
Will try again. You are also using bitcoin to control the TOR instance though.

**Comment by NicolasDorier** [comment]
It has been solved by adding `bind=127.0.0.1:18333` in config file (for testnet, for mainnet, use the default mainnet port 8333)
